### PR TITLE
[BugFix] use default avatar for ghost user (fix 500 error)

### LINKED ISSUE #8853: Deleted user's (Ghost) avater image generates error 500
- Gitea version (or commit ref): 1.10-rc2
- Git version: 2.20.1
- Operating system: Debian 10
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL (mariadb)
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [X] No (READ DESCRIPTION)
  - [ ] Not relevant
- Log gist:

`
2019/11/06 08:19:24 ...uters/user/avatar.go:29:Avatar() [E] Requested avatar for invalid user: user does not exist [uid: 0, name: Ghost, keyid: 0]
`

## Description

When a user, who created an issue, deletes itself, the issue is re-assigned to "Ghost", who does not have an image.

https://codeberg.org/user/avatar/Ghost/-1 generates error 500

See here:

https://codeberg.org/crimeflare/cloudflare-tor/issues/16

This does not happen on try.gitea.io because someone created a Ghost user there when it was not yet blacklisted. So the avatar is there.

https://try.gitea.io/Ghost

### LINKED ISSUE #9189: template error in issues list when poster_id is -1
- Gitea version (or commit ref):  1.10.0+14-gc01afd584
- Git version: 2.20.1
- Operating system: Debian 10
- Database:
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist:

`
template: repo/issue/list:236:26: executing "repo/issue/list" at <.Poster.ID>: nil pointer evaluating *models.User.ID
`

## Description

Deleted owners issues sometimes get poster_id=-1 in issue table, leading to a template error in ./templates/repo/issue/list.tmpl in the issue list

This happened for 5 issues on codeberg.org, in an organization where users create one-time accounts for one single interaction, then delete it (I won't comment on that, not my idea), but that fluctuation seems to make the error much more likely.

I am willing to fix the template IF you tell me that -1 is expected, if it is not then we have to find out why poster_id gets set to -1. Or is it rather a bug that deleted users poster_id do NOT get to -1?

### PR DESCRIPTION
**Pull request by 6543**
fix #8853 (set default avatar for ghost user (dont make a 500 error))

### CODE REVIEW

**Comment by 6543** [comment]
@ashimokawa the avatar issue was on codeberge too - but i can't find the releated issue here on github :(

**Comment by 6543** [comment]
... found it: https://codeberg.org/Codeberg/Community/issues/93 -> #8853

**Comment by 6543** [comment]
I'll add test for #8853 ...

File: routers/routes/routes.go [hunk 1]
```diff
@@ -385,24 +385,6 @@ func RegisterRoutes(m *macaron.Macaron) {
 		m.Post("/keys/delete", userSetting.DeleteKey)
 		m.Get("/organization", userSetting.Organization)
 		m.Get("/repos", userSetting.Repos)
-
```
Comment (lunny) [inline routers/routes/routes.go]: These are unrelated.

Comment (6543) [reply routers/routes/routes.go]: @lunny so should i add this in a second PR?

**Comment by zeripath** [comment]
I'm not much of a fan of requerying the posterID before saving it but it's the simplest option.

Whilst we're doing this we should look through the places that update the posterID to prevent them from updating a posterID to -1. 

I would also say we should also double check code that is getting users to ensure that they're actually all coping with ghost properly. (I personally had forgotten about ghost and so I'm not sure that I've explicitly coped with it.)

**Comment by 6543** [comment]
@zeripath look at backport comment ...

**Comment by zeripath** [comment]
Yeah seen - fix loadPoster and make update log an error.

File: routers/routes/routes.go (hunk 1 above)
Comment (zeripath) [reply routers/routes/routes.go]: Yes another PR and reference the PR that moved them too.

I suspect I merged it but if we're going about moving endpoints - the pr that does the movement should put the redirects in. (As fun as 2ndry bug fix PRs are, and we've had a flurry of them recently - mostly my fault - we should endeavour to get this right first time.)

File: models/issue.go [hunk 2]
```diff
@@ -1569,6 +1569,12 @@ func SearchIssueIDsByKeyword(kw string, repoIDs []int64, limit, start int) (int6
 }
 
 func updateIssue(e Engine, issue *Issue) error {
+	if issue.PosterID <= 0 {
+		_, err := e.Table("`issue`").Where("id = ?", issue.ID).Select("poster_id").Get(&issue.PosterID)
```
Comment (zeripath) [inline models/issue.go]: Please stick a log.Error after this if the posterID becomes > -1 with a comment to say that the calling methods need to be fixed and suggest stack trace. You should include the issue.ID.

If the posterID is still negative - then I think we should consider a log.Critical as a posterID<0 means a data loss has occurred at some point. That error message will definitely require the issue.ID in the message. (I'm not sure about Critical here but it's a definitely an error and the server owner needs to know about it.)

File: routers/routes/routes.go (hunk 1 above)
Comment (6543) [reply routers/routes/routes.go]: @lunny @zacheryph moved: #9538

**Review by lafriks — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
ready to merge :rocket:

**Comment by lafriks** [comment]
Please send backport

**Comment by 6543** [comment]
@techknowlogick can you add lable **bakport done** please?
